File: 206_reverse_linked_list.go
package amazon

func reverseListIterative(head *ListNode) *ListNode {
	pseudoHead := &ListNode{}
	var next *ListNode
	for p := head; p != nil; {
		next = p.Next
		p.Next = pseudoHead.Next
		pseudoHead.Next = p
		p = next
	}
	return pseudoHead.Next
}

func reverseListRecursive(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	newHead := reverseListRecursive(head.Next)
	head.Next.Next = head
	head.Next = nil
	return newHead
}

File: archived/python_hot_100/206_reverse_linked_list.py
from typing import Optional, Any, Sequence, List
from header import *

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fake_header = ListNode(next=head)
        p = fake_header.next
        while p is not None:
            next = p.next
            print(next)
            p.next = fake_header.next.next
            fake_header.next = p
            print(p.val, p.next)
            p = next
            break
        return fake_header.next


if __name__ == '__main__':
    n = ListNode(val=2)
    head = ListNode(val=1)
    s = Solution()
    new_head = s.reverseList(head)


    p = new_head
    while p is not None:
        print(f"({p.val}, {id(p.next)}")
        p = p.next